### ISSUE TITLE
Unable to handle WebRTC reconnect with WebRTCMultiplayer

### ISSUE THREAD

**Issue by dsnopek** [godotengine/godot#33010]
**Godot version:**

Godot 3.2-alpha at commit https://github.com/godotengine/godot/commit/3564547208030264a28cc55232992aab3cbdad94

**OS/device including version:**

Linux X11 and HTML5

**Issue description:**

Under controlled conditions, my WebRTC peer connections are stable and don't disconnect. However, when trying over the internet in real playtesting, it's quite common for a single WebRTC peer connection to break. I'm able to see it happen by `connect()`ing to the 'peer_disconnected' signal on `WebRTCMultiplayer`.

On that signal, I can establish a new `WebRTCPeerConnection` over the signalling server (a Web socket connection, that frequently remains stable even if some of the WebRTC ones break) and add it again via `WebRTCMultiplayer.add_peer()` using the same peer id.

If this happens before sending any `rpc()` calls (ie. during the lobby phase), then everything works fine! However, if any `rpc()` calls have been made (ie. when the match has already begun), then further `rpc()` calls after adding the new (reconnected) peer won't work and I get these messages:

```
ERROR: _process_packet: Invalid packet received. Requested node was not found.
   At: core/io/multiplayer_api.cpp:202.
ERROR: _process_get_node: Invalid packet received. Unabled to find requested cached node.
   At: core/io/multiplayer_api.cpp:263.
```

I'm not sure at all how to work around that. Given that WebRTC connections breaking and needing to be reconnected seems like a normal thing (per my research) and the fact that `WebRTCMultiplayer` has `.add_peer()` and `.remove_peer()` methods, I would think that it should be able to handle this? Or am I handling reconnect totally wrong?

Beyond that, in reading about reconnecting WebRTC connections in the context of the browser, they mention that the "disconnected" status is considered temporary, and that the connection may repair itself, and it's the "failed" status where you should attempt a reconnect. However, in `WebRTCMultiplayer::poll()` it's treating both "disconnected" and "failed" the same, and removing the peer and emitting the 'peer_disconnected' signal. Shouldn't it wait until "failed" because a "disconnected" connection could repair itself?

Pinging @Faless 

**Steps to reproduce:**

Unfortunately, WebRTC is complex with needing a signalling server, and this depends on real world networking conditions, so it'd be tricky to give steps to reproduce.

**Minimal reproduction project:**

Same for a minimal project, but I'll think on it. Maybe I can modify one of the example projects with a debug key to purposefully close one of the `WebRTCPeerConnection`.

**Comment by dsnopek** [comment]
Actually, I think I might have found a workaround! But it seems excessive...

If on reconnect, I disconnect all signals from the current `WebRTCMultiplayer`, create a new one and set via `get_tree().set_network_peer()`, add all the same `WebRTCPeerConnections` to it (except for the one that disconnected), and then add the new reconnected peer to it, then the game can continue.

I'm going to keep playtesting this for a few days and see how it goes. But if this does work, I still think this should be easier.

**Comment by Faless** [comment]
Judging from where the error comes from (`MultiplayerAPI`), it seems the problem is at the upper layer.
As a wild guess I would say it might be something about the path/node cache for the disconnecting peer not being cleaned up correctly.
It'll look into it, but it might take a while :( .

**Comment by dsnopek** [comment]
Yeah, I think it might be that `MultiplayerAPI` isn't prepared for a `NetworkMultiplayerPeer` that may remove and re-add peers on the fly. And, unfortunately, the way that `WebRTCMultiplayer` works, you kind of have to do that, because as soon as a `WebRTCPeerConnection` is disconnected (which happens all the time in my playtesting, and my research confirms is kind of normal for WebRTC), `WebRTCMultiplayer` will remove it.

If `WebRTCMultiplayer::poll()` didn't just remove it right away, but instead allowed your game to try reconnect it first, then maybe it wouldn't confuse `MultiplayerAPI`?

Anyway, thanks @Faless, for all your hard work on this! Despite this issue, my game is working pretty good in playtesting using WebRTC. :-) We've done a couple of matches where the players where all over the Americas (USA, Mexico and Colombia in the same match) and it was totally playable!

**Comment by Faless** [comment]
@dsnopek can you try this patch? I tested it under very different conditions, but I'm quite confident it's the right fix (I mean, we are obviously forgetting to clear that cache right now :) ).

```diff
diff --git a/core/io/multiplayer_api.cpp b/core/io/multiplayer_api.cpp
index 0ba84d0c8f..c14aae888c 100644
--- a/core/io/multiplayer_api.cpp
+++ b/core/io/multiplayer_api.cpp
@@ -603,6 +603,13 @@ void MultiplayerAPI::_add_peer(int p_id) {
 void MultiplayerAPI::_del_peer(int p_id) {
        connected_peers.erase(p_id);
        path_get_cache.erase(p_id); // I no longer need your cache, sorry.
+       // Remove peer from sent cache (TODO optimize?)
+       List<NodePath> keys;
+       path_send_cache.get_key_list(&keys);
+       for (List<NodePath>::Element *E = keys.front(); E; E = E->next()) {
+               PathSentCache *psc = path_send_cache.getptr(E->get());
+               psc->confirmed_peers.erase(p_id);
+       }
        emit_signal("network_peer_disconnected", p_id);
 }

```

**Comment by dsnopek** [comment]
@Faless Your patch works!

I added a debug key to my game to randomly close a WebRTCPeerConnection so that I could test that the game can reconnect them. With the patch in the previous comment, I can remove my workaround, and the game can still continue after reconnect! I'll start doing some real world testing too over the next couple of days.

Anyway, thanks so much! If this fix can get in before 3.2 comes out, it'll make WebRTC easier to manage for everyone. :-)

> That is very nice to hear! I'll have to try your game when it's out

It's an Open Source game so you can check it out in its current state if you want. I wrote a blog post yesterday about the conversion to WebRTC (which, of course, includes a link to play it):

https://www.snopekgames.com/blog/2019/switched-webrtc-networking-and-changed-name-retro-tank-party
